### ISSUE TITLE
[thold 1.0.4] Time in human readable data and option to select Graph time window (Enhancement)

### ISSUE THREAD

**Issue by g1augusto** [Cacti/plugin_thold#136]
It would be very useful if the <TIME> Variable would be in a human readable data instead of **Epoch** format, our graphs are in UTC.

Also the time window for the attached graph seems that cannot be selected and it can be confusing if more alerts are coming through (1 day default is too much I think)

Thanks

**Comment by g1augusto** [comment]
ADD:

How to use the threshold notes ? the <NOTES> variable in building the email refers to the device notes, not to the threshold notes :(

**Comment by cigamit** [comment]
Screen shot please.  Also the &lt;NOTES&gt; issue is resolved.

**Comment by netniV** [comment]
@cigamit I think he's refering to the fact that most tables within Cacti Core or Cacti Plugins use timestamp in the datebase.  That means if you are doing a select statement, as a human, it means very little compared to most modern databases which actually store as a datetime.

I think the safest way to have that handled would be to add a new field which was named _datetime (or something similar) which was automatically updated with the datetime (can that be auto updated but readonly to UPDATE/REPLACE statements, ie SQL Caculated Field?)

**Comment by g1augusto** [comment]
Thanks for your reply,

I set up some alerts for threshold breach on a graph for IP SLA (yes always that one :) ), and here you can see what I mentioned in my previous post:

- timestamp of the event is in epoch format, not human readable
- the graph that is shown shows 24 hours time, this is misleading as you can see on that graph we have several breach and it's unclear to which one refers to, it would be better to have an option to select the time window for the graph inserted into the email notification, like -30 minutes or -10 minutes


![image](https://user-images.githubusercontent.com/26431888/36475823-3bcd5180-16fc-11e8-8459-18a31f57c048.png)

**Comment by netniV** [comment]
Ah, that should be fixable... one sec...

**Comment by netniV** [comment]
Err, actually, might be slightly harder, can you provide the template text you are using?  I think it's a templated field.

**Comment by cigamit** [comment]
@netniV, we are using timestamps due to the fact that they store and regurgitate information in a zone dependent manor.  So, if I'm say in another country that is in another time zone, and as application Cacti knows this (we are not doing this anywhere today), it can represent the date correctly to the user by just switching the zone setting in the database session.  I know a few plugins that use this functionality today.  It's quite helpful.  In this case, it's just the note processing in thold_functions.php.  Not a big deal.

**Comment by netniV** [comment]
Yeah, so pretty much all timestamps are UTC, just those pesky date fields eh? ;-)

My earlier post was more for the SQL friendly people, so they can see the fields, like if your using phpMyAdmin.  Is there a computed field equivalent in MySQL?

**Comment by cigamit** [comment]
The update on the README.md should be pretty self explanatory.  I've added &lt;DATE&gt;, and documented &lt;DATE_RFC822&gt; to the threshold messages, and added &lt;TIME&gt;, &lt;DATE&gt;,&lt;DATE_RFC822&gt; to device Emails.  There is more cleanup to do.  I performed an analysis of all text replacement logic, and there are still some gaps.  But this will do for now.

**Comment by cigamit** [comment]
The timestamp will be stored always as UTC on disk, but output using the SELECT column FROM ... in the users timezone.  You can use format characters if you like, but otherwise it's Y-M-D H:I:S.

**Comment by cigamit** [comment]
@jhonnyx82, close if satisfied.  Change all &lt;TIME&gt; to &lt;DATE;&gt; and you will be set.

**Comment by g1augusto** [comment]
Thank you, 

I will have time tomorrow to test this, can I ask if also the time window in the alert graph can be customized or it will be added (eventually) in another update?

**Comment by g1augusto** [comment]
A question:

How can I update the thold plugin without losing the configuration of the thresholds ?

I actually overwrote the directory with the new develop extracted from a downloaded zip but I lost the previous settings, it wasn't a problem as we are still testing it but for the future in production it would be good to do it the right way.

Can you give me some guidance?

**Comment by netniV** [comment]
To be honest, I use git for all my updating.  I don't use the zip files, as it's cleaner easier etc.  However, for plugins it means you tend to always get the latest development code as they aren't marking plugins with release tags  at the moment.

As long as you are putting the sources in the same directory along with the INFO file, you should not loose ANY settings.  Settings are in the DB not the sources.

On the git method, I even use this to push out changes to a datacentre I've been doing some work for as in a couple of commands I mark it ready for them via a git push on their branch, login, git pull, and away they go.

**Comment by g1augusto** [comment]
most probably it's in the way I performed the update: 

1. Console -> Plugin Management
2. Disable plugin
3. Uninstall plugin
4. unzip thold develop into **plugin_thold_develop** folder
5. change ownership to proper user and group of **plugin_thold_develop** files
6. overwrite the **thold** plugin cacti folder
7. Console -> Plugin Management
8. install plugin
9. enable plugin

**Comment by netniV** [comment]
Don't uninstall!  Disable, by all means but DON'T UNINSTALL ... if you do that any plugin will normally remove ALL tables/columns it creates and thus wipe your configuration.  Really, there should be a warning about that... time for another issue 🗡

**Comment by g1augusto** [comment]
:) 

Thought about it

OK since we are here...I noticed that all data from the main poller is usually replicated to the remote pollers

Does this include the cacti install folder?
if yes does it applies to the plugins too?

I mean is that the case if I update the main poller, does it replicate the files to all remote pollers by design including the plugins?

until now I updated each poller manually after the main poller, or better at the same time (multiple SecureCRT sessions in parallel).

Any suggestion for this ?

**Comment by netniV** [comment]
Errr, no idea.

**Comment by g1augusto** [comment]
so strange

Another weird thing that is happening is that the plugin is getting disabled by itself :(



```
22/Feb/2018 10:34:28 - CMDPHP PHP ERROR NOTICE Backtrace: (/settings.php: 216 CactiErrorHandler)(/lib/functions.php: 4524 cacti_debug_backtrace)
--
22/Feb/2018 10:34:28 - ERROR PHP NOTICE: Undefined index: alerts in file: /var/www/html/cacti/settings.php on line: 216
22/Feb/2018 10:30:17 - CMDPHP ERRORS DETECTED - DISABLING PLUGIN 'thold'
22/Feb/2018 10:30:17 - CMDPHP PHP ERROR Backtrace: (CactiShutdownHandler)(/lib/functions.php: 4562 cacti_debug_backtrace)
22/Feb/2018 10:30:17 - ERROR PHP ERROR in Plugin 'thold': Call to undefined function thold_replace_tags() in file: /var/www/html/cacti/plugins/thold/thold_functions.php on line: 2789

```

I am still running version 1.1.34, can that be the case?

**Comment by netniV** [comment]
```diff
diff --git a/thold_functions.php b/thold_functions.php
index 7e7bf85..c8a8abb 100644
--- a/thold_functions.php
+++ b/thold_functions.php
@@ -2786,7 +2786,7 @@ function get_thold_alert_text($data_source_name, $thold, $h, $currentval, $local
        }

        if ($thold['notes'] != '') {
-               $notes = thold_replace_tags($thold['notes'], $thold, $h, $currentval, $local_graph_id, $data_source_name);
+               $notes = thold_replace_threshold_tags($thold['notes'], $thold, $h, $currentval, $local_graph_id, $data_source_name);
                $alert_text = str_replace('<NOTES>', $notes, $alert_text);
        }

@@ -2807,7 +2807,7 @@ function get_thold_warning_text($data_source_name, $thold, $h, $currentval, $loc
        }

        if ($thold['notes'] != '') {
-               $notes = thold_replace_tags($thold['notes'], $thold, $h, $currentval, $local_graph_id, $data_source_name);
+               $notes = thold_replace_threshold_tags($thold['notes'], $thold, $h, $currentval, $local_graph_id, $data_source_name);
                $warning_text = str_replace('<NOTES>', $notes, $warning_text);
        }

```

**Comment by g1augusto** [comment]
OK,

I remanaged the plugin folder, but at a certain point I got this error (I think in the case that a threshold alert should have been triggered)


```
22/Feb/2018 10:45:17 - CMDPHP SQL Backtrace: (/poller.php: 722 api_plugin_hook)(/lib/plugins.php: 73 api_plugin_run_plugin_hook)(/lib/plugins.php: 170 thold_poller_bottom)(/plugins/thold/includes/polling.php: 40 thold_check_all_thresholds)(/plugins/thold/includes/polling.php: 458 thold_check_threshold)(/plugins/thold/thold_functions.php: 2070 db_execute_prepared)(/lib/database.php: 208 cacti_debug_backtrace)
--
22/Feb/2018 10:45:17 - CMDPHP ERROR: A DB Exec Failed!, Error: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'thold_warning_fail_count = '1' thold_fail_count = 0 WHERE id = '2'' at line 1
22/Feb/2018 10:45:17 - DBCALL ERROR: A DB Exec Failed!, Error:1064, SQL:'UPDATE thold_data SET thold_alert = ? thold_warning_fail_count = ? thold_fail_count = 0 WHERE id = ?'
```

**Comment by netniV** [comment]
There was a problem with my above difference.  Just match sure you didn't apply `thold_replace_threshold_tags` as `thold_replace_replace_tags`

**Comment by netniV** [comment]
There appears to be a comma missing between thold_fail_count and thold_warning_fail_count

**Comment by netniV** [comment]
```diff
--- a/thold_functions.php
+++ b/thold_functions.php
@@ -2063,8 +2063,8 @@ function thold_check_threshold(&$thold_data) {
                        }

                        db_execute_prepared('UPDATE thold_data
-                               SET thold_alert = ?
-                               thold_warning_fail_count = ?
+                               SET thold_alert = ?,
+                               thold_warning_fail_count = ?,
                                thold_fail_count = 0
                                WHERE id = ?',
                                array($thold_data['thold_alert'], $thold_data['thold_warning_fail_count'], $thold_data['id']));
```

**Comment by g1augusto** [comment]
Oh, I didn't expect that.

How comes other users didn't see this coming when applying the new thold code ? :/

**Comment by netniV** [comment]
So that only occurs when you are using warnings up/down.  Most people are probably using alerts over warnings.

**Comment by netniV** [comment]
Commit bbddc94c360ebacd291ff02302f962f159b3c5e0 which updated a bunch of formatting on the SQL introduced the bug were the commas were missed switching from an SQL string to an SQL prepared statement (more secure)

**Comment by g1augusto** [comment]
ok, testing

**Comment by cigamit** [comment]
Sorry about all the mistakes guys.  Working too many hours.

**Comment by g1augusto** [comment]
> Sorry about all the mistakes guys. Working too many hours.

Don't even say that :)

**Comment by g1augusto** [comment]
UP/DOWN devices notification looks to be ok, I tested with the Global list, not with custom notification list yet but I think it will be fine.

About Threshold notification I am still testing

**Comment by netniV** [comment]
> Sorry about all the mistakes guys. Working too many hours.

That's been me over the past week too.  Normally why we catch each other though lol

**Comment by g1augusto** [comment]
**Issue 1** : Creating a Threshold from a graph with the "Create Threshold" button doesn't work, just returns a blank screen, no logs recorded, the URL called is : 
`https://xxx/cacti/plugins/thold/thold.php?action=add&host_id=&local_graph_id=41196
`
(Maybe host id is missing ?)

**Issue 2** : I think that the thresholds are not anymore referencing the graph data, I tried creating a new threshold alert value but I can see that the "current" value in the threshold data is not taken from the actual "current" value of the graph data and the threshold is not triggered: 

![image](https://user-images.githubusercontent.com/26431888/36538779-f7077dda-17d4-11e8-9b2d-e557a9233b14.png)

![image](https://user-images.githubusercontent.com/26431888/36538826-15ef2a9a-17d5-11e8-91f8-9ae392633bec.png)

**Comment by netniV** [comment]
I'm not sure how you went about that, can you give me the exact steps so I can retrace. I may be looking in the wrong place.

**Comment by g1augusto** [comment]
For **issue 1**

1. Browse to any graph under the main **Graph tab** in CACTI
2. Look at a graph and having the Thold plugin enabled you will see a button on the right side stating : Create threshold  ![image](https://user-images.githubusercontent.com/26431888/36539696-f370462c-17d7-11e8-8236-ca852958e871.png)
3. in the next menu select **create new threshold**
4. now you have a white empty screen :) 

For **issue 2**

1. Once you create the threshold (currently by Console -->Thresholds)
2. select a device - a graph - a data source 
3. set some values ( in my case I have some loss on an IP SLA, so it's lossDS and lossSD) I set a value of 1 high for both warning and alert, with high/low type threshold, this should trigger an alert if it goes above 1

From the graph I attached, the current value should be taken from the related graph at each polling interval (in my case 5 minutes) and it's not there, as you can see the threshold is 1 high but at 2.95 it wasn't triggered

**Comment by netniV** [comment]
Which mode are you in, tree view, preview (I'm presuming not list)

**Comment by g1augusto** [comment]
tree view

**Comment by netniV** [comment]
Does the link have usetemplate=1 ?

**Comment by g1augusto** [comment]
This is the link of the button:

`https://xxxx/cacti/plugins/thold/thold.php?action=add&usetemplate=1&local_graph_id=41196`

**Comment by netniV** [comment]
Thought so, so I know where to start from now.

**Comment by netniV** [comment]
So, here are the steps:

- call thold.php 
   - because there is a local_graph_id and no host_id, the code reads in the host id.  
   - because you have usetemplate but no thold_template_id, it calls thold_add_graphs_action_prepare()
      - thold_add_graphs_action_prepare() uses local_graph_id to read the host_id from graph_local
      - thold_add_graphs_action_prepare() then starts to output data.

So, do you have any errors in your logs when you try this action because you should have had some output.  Do you have the 'Create Threshold from Template' page and is it from that  you get no response?

**Comment by g1augusto** [comment]
`Do you have the 'Create Threshold from Template' page and is it from that you get no response?`
Correct

`do you have any errors in your logs when you try this action`
No logs :(

**Comment by netniV** [comment]
Try this patch:
```diff
diff --git a/thold_webapi.php b/thold_webapi.php
index d46803a..3fc9393 100644
--- a/thold_webapi.php
+++ b/thold_webapi.php
@@ -291,6 +291,10 @@ function thold_add_graphs_action_prepare() {
                                'method' => 'hidden',
                                'value' => 1
                        ),
+                       'host_id' => array(
+                               'method' => 'hidden',
+                               'value' => $host_id
+                       ),
                        'local_graph_id' => array(
                                'method' => 'hidden',
                                'value' => $local_graph_id
```

**Comment by g1augusto** [comment]
Sorry, same result

**Comment by netniV** [comment]
OK, so after you click the little button, and you get the confirmation screen, can you view the source and see if host_id is in there.

**Comment by netniV** [comment]
I'm going to guess, you are getting this error in the console of the browser:
```js
Uncaught ReferenceError: noFileSelected is not defined
    at themeReady (main.js?9750806ae1aa971056ea5cd9400148a4:44)
    at applySkin (layout.js?5a049141323c7a5933a44117ff5f7044:597)
    at HTMLDocument.<anonymous> (layout.js?5a049141323c7a5933a44117ff5f7044:1750)
    at mightThrow (jquery.js?9c3a8d5bf79a2b2c25b4d9f99fbf6db2:3583)
    at process (jquery.js?9c3a8d5bf79a2b2c25b4d9f99fbf6db2:3651)
```

**Comment by netniV** [comment]
Try this, it's a full diff from the repo source so you'll have to undo the change above by taking out the host_id array
```diff
diff --git a/thold_webapi.php b/thold_webapi.php
index d46803a..0a299e7 100644
--- a/thold_webapi.php
+++ b/thold_webapi.php
@@ -291,6 +291,10 @@ function thold_add_graphs_action_prepare() {
                                'method' => 'hidden',
                                'value' => 1
                        ),
+                       'host_id' => array(
+                               'method' => 'hidden',
+                               'value' => $host_id
+                       ),
                        'local_graph_id' => array(
                                'method' => 'hidden',
                                'value' => $local_graph_id
@@ -351,19 +355,16 @@ function thold_add_graphs_action_prepare() {
                                $.post(strURL, json).done(function(data) {
                                        document.location = '<?php print $backto;?>';
                                });
-                       }else if ($('#doaction').length) {
-                               strURL += (strURL.indexOf('?') >- 0 ? '&':'?');
-                               strURL += 'doaction='+$('#doaction').val();
-                               strURL += '&local_graph_id='+$('#local_graph_id').val();
-                               strURL += '&host_id='+$('#host_id').val();
-                               strURL += '&usetemplate=1'+$('#usetemplate').val();
-                               document.location = strURL;
-                       }else{
+                       }else {
                                strURL += (strURL.indexOf('?') >- 0 ? '&':'?');
-                               strURL += 'thold_template_id='+$('#thold_template_id').val();
                                strURL += '&local_graph_id='+$('#local_graph_id').val();
                                strURL += '&host_id='+$('#host_id').val();
-                               strURL += '&usetemplate=1'+$('#usetemplate').val();
+                               strURL += '&usetemplate='+$('#usetemplate').val();
+                               if ($('#doaction').length) {
+                                       strURL += '&doaction='+$('#doaction').val();
+                               } else {
+                                       strURL += '&thold_template_id='+$('#thold_template_id').val();
+                               }
                                document.location = strURL;
                        }
                });
```

**Comment by netniV** [comment]
Just tweaked the diff slightly.

**Comment by netniV** [comment]
OK, so despite the above fixes, the problem still persists that you get a blank screen.  The problem appears to be with top_header() and I just noticed that none of these functions are passing the header=false flag around.

**Comment by g1augusto** [comment]
do you need me to try it anyway or do you want me to test the next code ?

**Comment by netniV** [comment]
That pull request should resolve your issue https://github.com/Cacti/plugin_thold/pull/166

**Comment by netniV** [comment]
@jhonnyx82 can you try the code from the pull request?

**Comment by g1augusto** [comment]
I just tried and it blocks on the CREATE button

![image](https://user-images.githubusercontent.com/26431888/36554291-26c08290-17ff-11e8-8306-c2c3a8daa160.png)

**Comment by g1augusto** [comment]
EDIT: give me a moment as it seems it was due to the fact that it was not yet propagated the file on the remote poller

**Comment by netniV** [comment]
Hey, I just fixed the blank screen... that’s a new issue ha

**Comment by g1augusto** [comment]
It seems OK now, but did you commit another change? I used the #166 code

**Comment by g1augusto** [comment]
Now the problem still pending is the fact that the thresholds configured are not evaluating the actual data source data

**Comment by netniV** [comment]
Thing is, this thread has got wildly off course with other issues... Do you have monitor installed? Does it show the thresholds as breached or triggered?

**Comment by g1augusto** [comment]
It doesn't read the current status of the Threshold at all:

![image](https://user-images.githubusercontent.com/26431888/36556173-8a3b797e-1804-11e8-8efd-745e34fff04a.png)

![image](https://user-images.githubusercontent.com/26431888/36556179-93775d5a-1804-11e8-9bbc-759edd35c877.png)

**Comment by g1augusto** [comment]
ADD (This is related still to the threshold creation blank screens issue):

on some data source graphs I am getting this error while trying to create the tresholds:


```
22/Feb/2018 18:16:03 - CMDPHP PHP ERROR NOTICE Backtrace: (/plugins/thold/thold.php: 64 save_thold)(/plugins/thold/thold_functions.php: 3603 CactiErrorHandler)(/lib/functions.php: 4524 cacti_debug_backtrace)
--
22/Feb/2018 18:16:03 - ERROR PHP NOTICE in Plugin 'thold': Undefined index: graph_template_id in file: /var/www/html/cacti/plugins/thold/thold_functions.php on line: 3603
22/Feb/2018 18:16:03 - CMDPHP PHP ERROR NOTICE Backtrace: (/plugins/thold/thold.php: 64 save_thold)(/plugins/thold/thold_functions.php: 3602 CactiErrorHandler)(/lib/functions.php: 4524 cacti_debug_backtrace)
22/Feb/2018 18:16:03 - ERROR PHP NOTICE in Plugin 'thold': Undefined index: local_graph_id in file: /var/www/html/cacti/plugins/thold/thold_functions.php on line: 3602
```

**Comment by netniV** [comment]
Hmm, I'll take a look.

**Comment by netniV** [comment]
That error occurs when saving a threshold.  Also, your line numbers do not match up with mine so I'm not sure what's occurring there.

**Comment by cigamit** [comment]
I'm going to close this one as the original feature enhancement has been addressed.  I have not been able to duplicate the last issue here using the latest code.  But, if it continues, please open a new issue stating the specific steps you took.  Screen shot per click please.
